modules/siddhi-extensions/markov-models/src/main/java/org/wso2/siddhi/extension/markov/MarkovChainTransitionProbabilitiesCalculator.java

```diff
@@ -0,0 +1,237 @@
+     */
+    protected void removeExpiredEvents(long currentEventTime) {
+
+        for (Map.Entry<String, Long> entry : idToLastStatesExpiryTime.entrySet()) {
+            if (entry.getValue() <= currentEventTime) {
+                idToLastStates.remove(entry.getKey());
+            } else {
+                break;
+            }
+        }
+    }
+
+    /**
+     * Method to populate the markov matrix using a file input
+     */
+    private void populateMarkovMatrix(String markovMatrixStorageLocation) {
+
+        File file = new File(markovMatrixStorageLocation);
+        FileInputStream fis = null;
+        BufferedInputStream bis = null;
+        BufferedReader br = null;
+        String[] statesNames = null;
+        String key;
+        String startState;
+        String endState;
+
+        Map<String, Double> transitionCount = new HashMap<String, Double>();
+        Map<String, Double> startStateCount = new HashMap<String, Double>();
+
+        try {
+
+            fis = new FileInputStream(file);
+            bis = new BufferedInputStream(fis);
+            br = new BufferedReader(new InputStreamReader(bis));
+            int rowNumber = 0;
+            String row = br.readLine();
+            if (row != null) {
+                statesNames = row.split(",");
+            }
+
+            while ((row = br.readLine()) != null) {
+
+                if (rowNumber >= statesNames.length) {
+                    throw new OperationNotSupportedException(
+                            "Number of rows in the matrix should be equal to number of states. please provide a "
+                                    + statesNames.length + " x " + statesNames.length + " matrix");
+                }
+
+                startState = statesNames[rowNumber];
+                String[] values = row.split(",");
+                double totalCount = 0.0;
+
+                if (values.length != statesNames.length) {
+                    throw new OperationNotSupportedException(
+                            "Number of columns in the matrix should be equal to number of states. please provide a "
+                                    + statesNames.length + " x " + statesNames.length + " matrix");
+                }
+
+                for (String value : values) {
+                    totalCount = totalCount + Double.parseDouble(value);
+                }
+                startStateCount.put(startState, totalCount);
+
+                for (int j = 0; j < statesNames.length; j++) {
+                    endState = statesNames[j];
+                    key = markovMatrix.getKey(startState, endState);
+                    transitionCount.put(key, Double.parseDouble(values[j]));
+                }
+                rowNumber++;
+            }
+
+        } catch (IOException e) {
+            log.error("Exception occurred while reading the data file " + markovMatrixStorageLocation, e);
+        } finally {
+            try {
+                if (fis != null) {
```

Comment:
Check, bis and fis closing order. General rule, should close streams in the reverse order we open.
Also handle IOException while closing the streams independently. We do not want not to close "br" when exception occurred while closing "fis".
Better have proper names for br, fis, bis.